Suggest std::ptr::null if literal 0 is given to a raw pointer function argument

edward-shen opened the pull request:
Implementation feels a little sus (we're parsing the span for a `0`) but it seems to fall in line the string-expected-found-char condition right above this check, so I think it's fine.

Feedback appreciated on help text? I think it's consistent but it does sound a little awkward maybe?

Fixes #107517

Nilstrieb commented on compiler/rustc_infer/src/infer/error_reporting/mod.rs:
```diff
@@ -1999,6 +1999,19 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
                             self.suggest_let_for_letchains(&mut err, &trace.cause, span);
                         }
+                        // Suggest providing a `std::ptr::null()` if they pass in a 0 to an *const uint
+                        (ty::RawPtr(ptr), ty::Uint(_)) if ptr.mutbl == rustc_hir::Mutability::Not => {
```
You can also suggest `std::ptr::null_mut()` for `*mut T`.

Nilstrieb replied:
```suggestion
                        (ty::RawPtr(ptr), ty::Uint(_)) if ptr.mutbl == hir::Mutability::Not => {
```

Nilstrieb commented on tests/ui/suggest-null-ptr.rs:
```diff
@@ -0,0 +1,22 @@
+// Suggest providing a std::ptr::null if provided with
```
```suggestion
// run-rustfix

// Suggest providing a std::ptr::null if provided with
```
To make sure that the suggestion works as expected.

edward-shen commented:
Changes:
 - Added `run-runfix` directive and a `.fixed` file 
 - `s/rustc_hir/hir/`
 - Added suggestion for `_mut() variant.

WaffleLapkin commented on compiler/rustc_infer/src/infer/error_reporting/mod.rs:
```diff
@@ -1999,6 +1999,31 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
+                                            err.span_suggestion(
+                                                span,
+                                                "if you meant to create a null pointer, use `std::ptr::null()`",
+                                                "std::ptr::null()",
```
Hard-coding `std` may be annoying for `#![no_std]` crates, please get the path to the diagnostic item instead.

The only way I've found so far that does that is the following:
```rust
let null_sym = match ptr.mutbl {
    hir::Mutability::Not => sym::ptr_null,
    hir::Mutability::Mut => sym::ptr_null_mut,
};

if let Some(null_did) = self.tcx.get_diagnostic_item(null_sym) {
    if let Ok((printer, true)) = FmtPrinter::new(self.tcx, Namespace::ValueNS).try_print_visible_def_path(null_did) {
        let path = printer.into_buffer();
        err.span_suggestion(
            span,
            format!("if you meant to create a null pointer, use `{path}`"),
            path + "()",
            Applicability::MachineApplicable,
        );
    }
}
```

compiler-errors replied:
all that `FmtPrinter` stuff might be avoided with a call to `with_no_trimmed_paths!(tcx.def_path_str(null_def_id))`

compiler-errors replied:
Also, this logic may better live in `FnCtxt::emit_type_mismatch_suggestions` (well, a method called from it). At that point, we should have access to the HIR expression that causes this mismatch. Then we can just check if it's a literal. 

Otherwise, I would personally not want to land this if we're doing that span_to_snippet + parsing. It seems pretty fragile.

compiler-errors replied:
Alternatively, we could probably make `report_and_explain_type_error` to take an optional `Option<&'tcx hir::Expr<'tcx>>`, and use that here, bubbling it down from hir typeck. That's a much more intrusive change, though.

eggyal commented on tests/ui/suggest-null-ptr.rs:
```diff
@@ -0,0 +1,31 @@
+}
+
+fn main() {
+    unsafe {
```
Nit: not required

workingjubilee commented:
`c_char` is understood to mean `i8` in some cases, it is surprising this suggestion would be limited to `u8`, etc.

Nilstrieb commented:
The diagnostic fires for raw pointers no matter the type behind it.

edward-shen commented on compiler/rustc_infer/src/infer/error_reporting/mod.rs:
```diff
@@ -1999,6 +1999,31 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
+                        // Suggest providing a std::ptr::null{,_mut}() if they pass in a 0 to an *{const,mut} uint
+                        (ty::RawPtr(ptr), ty::Uint(_))  => {
+                            if let Ok(code) = self.tcx.sess().source_map().span_to_snippet(span) {
+                                if let Ok(0) = code.parse::<u8>() {
```
Is additional string parsing a good idea here?

We can discard all `_`, followed by then taking the subset of the str that excludes anything after an `i` or `u`, if one was present. That should handle all cases I can think of (`000_000`, `0_0usize` `0_0_iu32`, and parse already supports something like `"0000000000000"` parsing as a valid `u8`). Maybe this work has already been done somewhere?

Alternatively we can just leave that as potential future work -- after all I don't think it will be common that folks will pass in a prefixed numeric literal to an raw pointer argument.

WaffleLapkin replied:
it's more that I don't think that parsing `span_to_snippet` is a good idea at all, see @compiler-errors' comments: https://github.com/rust-lang/rust/pull/107553#discussion_r1093963499

edward-shen commented:
Yeah, I think the commit name is better, let me update the PR name to better reflect the changes. Sorry for the confusion.

workingjubilee commented:
![ferris clueless](https://user-images.githubusercontent.com/46493976/216411420-d4649784-7012-4175-acab-3ae990e0d0cc.png)
Ope, my bad.

edward-shen committed:
Suggest null ptr if 0 is given as a raw ptr arg

edward-shen commented:
Changes:
 - Moved logic into a new function that is called by `FnCtxt::emit_type_mismatch_suggestions`
 - Cleaned up logic now that we have information about the HIR
 - Used `with_no_trimmed_paths` to print the diagnostic item instead.

WaffleLapkin reviewed (approved):
Thanks!

WaffleLapkin commented:
@bors r+

bors commented:
:pushpin: Commit 32967296b444b639c965cbf126720c24decc9c65 has been approved by `WaffleLapkin`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=WaffleLapkin 32967296b444b639c965cbf126720c24decc9c65 -->
<!-- homu: {"type":"Approved","sha":"32967296b444b639c965cbf126720c24decc9c65","approver":"WaffleLapkin","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
